### PR TITLE
Restructured dsym processing to support better error reporting

### PR DESCRIPTION
**Pull request by mitsuhiko**
This is a squashed change of all the dsym related improvements
that aid error reporting and better reasoning about how errors
are happening in the symbolication process.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
One concern with this is that `report_error` logs too much for missing system symbols. We might not want to log that if we do not see a value in it.

File: CHANGES [hunk 1]
```diff
@@ -1,6 +1,9 @@
 Version 8.12 (Unreleased)
 -------------------------
 
+- restructured dSYM processing.  It now records errors in the symbolication process
+  more accurately and will report system errors them to the internal logger.
```
Comment (mattrobenolt) [inline CHANGES]: Is this still accurate? Also, this is horribly broken english.

**Review by mattrobenolt — APPROVED** [review]
:ok:

**Comment by mitsuhiko** [comment]
This was merged before.

**Pull request closed without merging**